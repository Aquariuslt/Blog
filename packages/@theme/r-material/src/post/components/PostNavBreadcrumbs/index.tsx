import * as React from 'react';
import { Link as RouterLink } from 'react-router-dom';
import { capitalize } from 'lodash';
import { createStyles, makeStyles, StyleRules, Theme } from '@material-ui/core/styles';
import Breadcrumbs from '@material-ui/core/Breadcrumbs';
import Link from '@material-ui/core/Link';
import Paper from '@material-ui/core/Paper';
import Grid from '@material-ui/core/Grid';

import styles from '../../../styles';

export interface PostNavBreadcrumbsProps {
  path: string;
}

interface BreadcrumbLink {
  label: string;
  to: string;
  activated: boolean;
}

const useStyles = makeStyles(
  (theme: Theme): StyleRules =>
    createStyles({
      root: {
        display: 'flex',
        flexWrap: 'wrap',
        justifyContent: 'center',
        overflow: 'hidden',
        backgroundColor: theme.palette.background.paper
      },
      breadcrumbs: {
        maxWidth: styles.content.maxWidth,
        padding: theme.spacing(1, 1)
      }
    })
);

/**
 * A breadcrumbs that support all paths
 * */
const PostNavBreadcrumbs: React.ComponentType<PostNavBreadcrumbsProps> = (
  props: PostNavBreadcrumbsProps
): JSX.Element => {
  const classes = useStyles();

  const linkPaths = props.path.split('/').filter((value): boolean => value !== '');

  const linkContents = linkPaths.map(
    (link, index): BreadcrumbLink => {
      return {
        label: capitalize(link),
        to: '/' + linkPaths.slice(0, index + 1).join('/'),
        activated: index === linkPaths.length - 1
      };
    }
  );

  const links = linkContents.map(
    (link): JSX.Element => (
      <Link key={link.to} component={RouterLink} to={link.to} color={link.activated ? 'textPrimary' : 'inherit'}>
        {link.label}
      </Link>
    )
  );

  return (
    <Grid container className={classes.root}>
      <Grid item xs={12} sm={12} md={12} lg={12} xl={12} className={classes.breadcrumbs}>
        <Paper elevation={0}>
          <Breadcrumbs component="nav" area-label="Breadcrumb">
            {links}
          </Breadcrumbs>
        </Paper>
      </Grid>
    </Grid>
  );
};

export default PostNavBreadcrumbs;
