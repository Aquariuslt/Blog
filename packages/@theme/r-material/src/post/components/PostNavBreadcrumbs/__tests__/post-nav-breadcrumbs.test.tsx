import * as React from 'react';
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';

import PostNavBreadcrumbs from '../';


Enzyme.configure({ adapter: new Adapter() });

describe('@theme/r-material: post/component: PostNavBreadcrumbs', (): void => {

  it('# should mount correctly', (): void => {

    const path = '/categories/some-category';
    const wrapper = Enzyme.shallow(<PostNavBreadcrumbs path={path}/>);
    expect(wrapper).toMatchSnapshot();
  });

  it('# should mount with long path and render last one as activated', (): void => {

    const path = '/posts/2019/06/05';
    const wrapper = Enzyme.shallow(<PostNavBreadcrumbs path={path}/>);
    expect(wrapper).toMatchSnapshot();
  });
});
