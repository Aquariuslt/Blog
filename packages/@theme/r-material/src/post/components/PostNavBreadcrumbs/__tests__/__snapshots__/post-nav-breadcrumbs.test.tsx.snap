// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@theme/r-material: post/component: PostNavBreadcrumbs # should mount correctly 1`] = `
<WithStyles(ForwardRef(Grid))
  className="makeStyles-root-1"
  container={true}
>
  <script
    dangerouslySetInnerHTML={
      Object {
        "__html": "{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"BreadcrumbList\\",\\"itemListElement\\":[{\\"@type\\":\\"ListItem\\",\\"position\\":1,\\"name\\":\\"Categories\\",\\"item\\":\\"localhost/categories\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":2,\\"name\\":\\"Some-category\\",\\"item\\":\\"localhost/categories/some-category\\"}]}",
      }
    }
    type="application/ld+json"
  />
  <WithStyles(ForwardRef(Grid))
    className="makeStyles-breadcrumbs-2"
    item={true}
    lg={12}
    md={12}
    sm={12}
    xl={12}
    xs={12}
  >
    <WithStyles(ForwardRef(Paper))
      elevation={0}
    >
      <WithStyles(ForwardRef(Breadcrumbs))
        area-label="Breadcrumb"
        component="nav"
      >
        <WithStyles(ForwardRef(Link))
          color="inherit"
          component={[Function]}
          key="/categories"
          to="/categories"
        >
          Categories
        </WithStyles(ForwardRef(Link))>
        <WithStyles(ForwardRef(Link))
          color="textPrimary"
          component={[Function]}
          key="/categories/some-category"
          to="/categories/some-category"
        >
          Some-category
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Breadcrumbs))>
    </WithStyles(ForwardRef(Paper))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;

exports[`@theme/r-material: post/component: PostNavBreadcrumbs # should mount with long path and render last one as activated 1`] = `
<WithStyles(ForwardRef(Grid))
  className="makeStyles-root-1"
  container={true}
>
  <script
    dangerouslySetInnerHTML={
      Object {
        "__html": "{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"BreadcrumbList\\",\\"itemListElement\\":[{\\"@type\\":\\"ListItem\\",\\"position\\":1,\\"name\\":\\"Posts\\",\\"item\\":\\"localhost/posts\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":2,\\"name\\":\\"2019\\",\\"item\\":\\"localhost/posts/2019\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":3,\\"name\\":\\"06\\",\\"item\\":\\"localhost/posts/2019/06\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":4,\\"name\\":\\"05\\",\\"item\\":\\"localhost/posts/2019/06/05\\"}]}",
      }
    }
    type="application/ld+json"
  />
  <WithStyles(ForwardRef(Grid))
    className="makeStyles-breadcrumbs-2"
    item={true}
    lg={12}
    md={12}
    sm={12}
    xl={12}
    xs={12}
  >
    <WithStyles(ForwardRef(Paper))
      elevation={0}
    >
      <WithStyles(ForwardRef(Breadcrumbs))
        area-label="Breadcrumb"
        component="nav"
      >
        <WithStyles(ForwardRef(Link))
          color="inherit"
          component={[Function]}
          key="/posts"
          to="/posts"
        >
          Posts
        </WithStyles(ForwardRef(Link))>
        <WithStyles(ForwardRef(Link))
          color="inherit"
          component={[Function]}
          key="/posts/2019"
          to="/posts/2019"
        >
          2019
        </WithStyles(ForwardRef(Link))>
        <WithStyles(ForwardRef(Link))
          color="inherit"
          component={[Function]}
          key="/posts/2019/06"
          to="/posts/2019/06"
        >
          06
        </WithStyles(ForwardRef(Link))>
        <WithStyles(ForwardRef(Link))
          color="textPrimary"
          component={[Function]}
          key="/posts/2019/06/05"
          to="/posts/2019/06/05"
        >
          05
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Breadcrumbs))>
    </WithStyles(ForwardRef(Paper))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
